test_runner: introduce NODE_TEST_WORKER_ID for improved concurrent te…

cu8code opened the pull request:
ref: #55842 
 
Added a new environment variable, `NODE_TEST_WORKER_ID`, which ranges from 1 to N when `--experimental-test-isolation=process` is enabled and defaults to 1 when `--experimental-test-isolation=none` is used.

Before merging, I want to add some tests but haven't come up with a good approach yet. Here's what I aim to test:  
- When `--experimental-test-isolation=process` is enabled, verify that `NODE_TEST_WORKER_ID` ranges from 1 to N.  
- When `--experimental-test-isolation=none` is used, ensure that `NODE_TEST_WORKER_ID` is set to 1.  

Any suggestions on how to create such tests would be greatly appreciated!

mcollina reviewed (approved):
lgtm

pmarchini reviewed (changes requested):
Hey @cu8code, thanks for the PR.
I'm requesting changes since we're still missing tests.
Regarding the implementation, LGTM.

Another important thing: to make this new feature available to users, we also need to update the documentation with this behavior 🚀

JakobJingleheimer commented:
Regarding the tests, I think you'll want to use `spawnPromisified`. See [test-cjs-esm-warn.js](https://github.com/nodejs/node/tree/main/test/es-module/test-cjs-esm-warn.js)

JakobJingleheimer commented:
I'm concerned this has too much cross-over with `threadId`, for which the lack of sequentiality guarantee is likely for a reason—sequentiality likely would have been trivial to facilitate. I don't know who within our collaborators has that context though (maybe Anna?).

If there is a good reason behind making `threadId` non-sequential, I suspect this will be subject to that same rationale.

If there is not a particular rationale for the non-sequentiality of `threadId`, perhaps it would be better to guarantee that instead of creating a quasi-redundant mechanism.

The problem described in the cited issue is well handled via `threadId` when a complementary design is applied (which, IMO, is better than the attempted design the OP is trying to facilitate—which may well have been necessary in the environment in which it was created).

If we're merely trying to facilitate migration, I think we should first consider that point and whether there is a way to ease that migration without redundancy.

Without seeing the OP's code, I can only imagine what it looks like, but I suspect perhaps less than 20% would need to be adjusted. Likely the biggest hindrance is knowledge, which can be provided (ex a Learn article).

cu8code commented:
@JakobJingleheimer Thanks for helping with the test creation!  
@pmarchini I've added the test and updated the documentation—please take a look when you have a chance.

cjihrig commented on doc/api/test.md:
```diff
@@ -466,6 +466,15 @@ each other in ways that are not possible when isolation is enabled. For example,
 
+## Environment Variables
+
+### `NODE_TEST_WORKER_ID`
```
This should be documented in `doc/api/cli.md` instead.

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/1.test.js:
```diff
@@ -0,0 +1,7 @@
+// This file tests the `NODE_TEST_WORKER_ID` feature.
+// The test logic is implemented in `test/parallel/test-runner-cli.js`.
```
Can you remove this line. If things get moved around, this will almost certainly be forgotten about.

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/1.test.js:
```diff
@@ -0,0 +1,7 @@
+
+test('test1', t => {
+    console.log('NODE_TEST_WORKER_ID', process.env.NODE_TEST_WORKER_ID)
+});
```
Please add a newline at the end of the file.

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/1.test.js:
```diff
@@ -0,0 +1,7 @@
+// This file tests the `NODE_TEST_WORKER_ID` feature.
```
Please rename the directory from `NODE_TEST_WORKER_ID` to `worker-id`.

pmarchini replied:
@cu8code this still needs to be addressed 😁

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/2.test.js:
```diff
@@ -0,0 +1,7 @@
+// This file tests the `NODE_TEST_WORKER_ID` feature.
```
The same comments apply to this file.

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/3.test.js:
```diff
@@ -0,0 +1,7 @@
+// This file tests the `NODE_TEST_WORKER_ID` feature.
```
The same comments apply to this file.

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/4.test.js:
```diff
@@ -0,0 +1,7 @@
+// This file tests the `NODE_TEST_WORKER_ID` feature.
```
I think three files is enough for this. I would drop this file.

cjihrig commented on test/parallel/test-runner-cli.js:
```diff
@@ -428,3 +428,31 @@ for (const isolation of ['none', 'process']) {
   assert.strictEqual(child.signal, null);
 }
+
+{
```
Please move these new tests to a new file - `test/parallel/test-runner-worker-id.js`.

cjihrig reviewed (commented):
Left a few comments. Have you verified that this works with watch mode (particularly watch mode + process isolation)?

cu8code commented:
@cjihrig I’ve made the updates you requested. Please take a moment to review them!

pmarchini commented on test/parallel/test-runner-worker-id.js:
```diff
@@ -0,0 +1,35 @@
+const fixtures = require('../common/fixtures');
+const testFixtures = fixtures.path('test-runner');
+
+{
```
Could you please implement the tests using `node:test` ? 
Here is an example : https://github.com/nodejs/node/blob/main/test/parallel/test-runner-run.mjs

cjihrig commented on test/fixtures/test-runner/NODE_TEST_WORKER_ID/1.test.js:
```diff
@@ -0,0 +1,5 @@
+const test = require('node:test');
```
It may have been missed among the other changes, but please rename this directory from `NODE_TEST_WORKER_ID` to something like `worker-id`.

pmarchini commented:
> > (particularly watch mode + process isolation)?
> 
> yes I have tested, its working fine! both for `process` and `none`

Hey @cu8code, regarding this: could you please also add some tests?

cjihrig commented on doc/api/cli.md:
```diff
@@ -2249,6 +2249,19 @@ Starts the Node.js command line test runner. This flag cannot be combined with
 See the documentation on [running tests from the command line][]
 for more details.
 
+### Environment Variable
```
This belongs in the existing environment variables section of the document. It should go immediately after the `NODE_TEST_CONTEXT` entry. It should match the formatting of the other environment variables in the document, and should not have its own header.

cu8code replied:
something like this ?

```md
### `NODE_TEST_CONTEXT=value`

If `value` equals `'child'`, test reporter options will be overridden and test
output will be sent to stdout in the TAP format. If any other value is provided,
Node.js makes no guarantees about the reporter format used or its stability.

### `NODE_TEST_WORKER_ID`

This environment variable is set by the test runner. When using process-level test
isolation, each worker process is assigned a unique ID, starting at `1`. This is
set to 1 for all tests when test isolation is disabled.

cjihrig replied:
Yep

cjihrig commented on doc/api/cli.md:
```diff
@@ -2249,6 +2249,19 @@ Starts the Node.js command line test runner. This flag cannot be combined with
+
+`NODE_TEST_WORKER_ID`:
+
+* Parallel Test Runs: If tests are executed in parallel (using the `--test` flag
+  alone or with the `--experimental-test-isolation=process` flag), this variable is assigned
+  a unique number ranging from `1` to `N`, where `N` represents the total number of test files being processed.
+
+* Isolated Test Runs: When tests are run in isolation (using the `--experimental-test-isolation=none` flag),
+  this variable is always set to `1`.
```
Maybe just adapt the docs from Jest:

```
This environment variable is set by the test runner. When using process-level test isolation, each worker process is assigned a unique ID, starting at `1`. This is set to 1 for all tests when test isolation is disabled.
```

cu8code commented:
:love_you_gesture: @cjihrig @pmarchini  review

cjihrig commented on test/parallel/test-runner-worker-id.js:
```diff
@@ -0,0 +1,100 @@
+});
+
+
+test('testing with isolation enabled in watch mode', async () => {
```
The watch mode tests added here have at least two problems:

1. They don't actually test watch mode beyond passing the `--watch` flag. To properly test watch mode, you have to trigger the restart functionality.
2. These tests will almost certainly be flaky in the CI. Waiting for a second, killing the child process, and expecting it to have run successfully probably works locally, but the CI machines can be very slow. I don't think we want to have a timeout at all.

I think these need to be refactored so that:
- The test runs once.
- Once the expected output is seen, a source file is modified (we don't want to modify the fixtures files directly either) to trigger a restart.
- The test runs again. We need the output to be the same both times.

cu8code commented:
I am facing this error when running using `./node --test test/parallel/test-runner-worker-id.js` `Warning: node:test run() is being called recursively within a test file` what is the best solution to avoid this, I have also implemented this test using async `child`, `run` form "node:test", but the error is persistent when we use the --experimental-test-isolation=process flag! below all the implementation create the same error

case 1
```js
  it('test NODE_TEST_WORKER_ID when --experimental-test-isolation=process', () => {
    const args = ['--test', '--experimental-test-isolation=process'];
    const child = spawnSync(process.execPath, args, { cwd: join(testFixtures, 'worker-id') });

    assert.strictEqual(child.stderr.toString(), '');
    const stdout = child.stdout.toString();

    assert.match(stdout, /NODE_TEST_WORKER_ID 1/);
    assert.match(stdout, /NODE_TEST_WORKER_ID 2/);
    assert.match(stdout, /NODE_TEST_WORKER_ID 3/);

    assert.strictEqual(child.status, 0);
  });
```
case with .run

```js
it('when --experimental-test-isolation=process', async () => {
    const args = ['--test', '--experimental-test-isolation=process'];
    const specReporter = new spec();
    
    const stream = await run({
      cwd: join(testFixtures, 'worker-id'),
      execArgv: args
    }).compose(specReporter).toArray();
    
    let resString = "";
    stream.map((buffer) => { 
      resString += buffer.toString();
    });
    
    const workerIdRegex = /NODE_TEST_WORKER_ID\s+\d+/g;
    
    const matches = resString.match(workerIdRegex);
    
    assert.ok(matches, 'No worker IDs found');
    assert.strictEqual(matches.length, 3, 'Expected 3 worker IDs');
    
    assert.match(resString, /NODE_TEST_WORKER_ID 1/, 'Worker ID 1 not found');
    assert.match(resString, /NODE_TEST_WORKER_ID 2/, 'Worker ID 2 not found');
    assert.match(resString, /NODE_TEST_WORKER_ID 3/, 'Worker ID 3 not found');
});
```

case with spawn

```js
test('testing with isolation enabled', () => {
  const args = ['--test', '--experimental-test-isolation=process', '--test-reporter=spec'];
  const child = spawn(process.execPath, args, { 
    cwd: join(testFixtures, 'worker-id'),
    stdio: ['pipe', 'pipe', 'pipe'] // Explicitly set stdio to pipe
  });
  
  let outputData = '';
  let errorData = '';
  
  child.stdout.setEncoding('utf8');
  child.stderr.setEncoding('utf8');
  
  child.stdout.on('data', (data) => {
    console.log('Captured stdout:', data);
    outputData += data;
  });
  
  child.stderr.on('data', (data) => {
    console.error('Captured stderr:', data);
    errorData += data;
  });
  
  child.on('close', (code) => {
    console.log('Process closed with code:', code);
    console.log('Total output:', outputData);
    console.log('Total errors:', errorData);
  });
});
```

I get the error;
```
 test NODE_TEST_WORKER_ID when --experimental-test-isolation=process (173.226822ms)
    AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
    + actual - expected
    
    + '(node:42186) Warning: node:test run() is being called recursively within a test file. skipping running files.\n' +
    +   '(Use `node --trace-warnings ...` to show where the warning was created)\n'
    - ''
    
        at TestContext.<anonymous> (file:///home/ankan/Documents/git/me/node/test/parallel/test-runner-worker-id.js:17:12)
        at Test.runInAsyncScope (node:async_hooks:211:14)
        at Test.run (node:internal/test_runner/test:931:25)
        at Test.start (node:internal/test_runner/test:829:17)
        at node:internal/test_runner/test:1308:71
        at node:internal/per_context/primordials:483:82
        at new Promise (<anonymous>)
        at new SafePromise (node:internal/per_context/primordials:451:29)
        at node:internal/per_context/primordials:483:9
        at Array.map (<anonymous>) {
      generatedMessage: true,
      code: 'ERR_ASSERTION',
      actual: '(node:42186) Warning: node:test run() is being called recursively within a test file. skipping running files.\n(Use `node --trace-warnings ...` to show where the warning was created)\n',
      expected: '',
      operator: 'strictEqual'
    }

```

rluvaton reviewed (approved):
LGTM

pmarchini commented:
Hey @cu8code, commit linter is failing, could you please fix it? ☺️

MoLow commented on doc/api/cli.md:
```diff
@@ -3306,6 +3306,12 @@ If `value` equals `'child'`, test reporter options will be overridden and test
 output will be sent to stdout in the TAP format. If any other value is provided,
 Node.js makes no guarantees about the reporter format used or its stability.
 
+### `NODE_TEST_WORKER_ID`
```
I'm not sure this is the correct place for this. this is a list of env vars node accepts, not env vars it exposes

MoLow commented on lib/internal/test_runner/runner.js:
```diff
@@ -724,8 +724,10 @@ function run(options = kEmptyObject) {
+      let workerId = 1;
       return SafePromiseAllSettledReturnVoid(testFiles, (path) => {
-        const subtest = runTestFile(path, filesWatcher, opts);
+        const subtest = runTestFile(path, filesWatcher, opts, workerId);
```
if I have 8 cpus and 16 tests, I would expect each worker to be reused twice, but this implementation will not reuse worker ids, I think that is misleading

cjihrig commented:
@cu8code are you still working on this? If not, we should close this PR so that other folks can pick up the work.
